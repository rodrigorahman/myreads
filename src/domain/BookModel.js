class BookModel {

    constructor(id, thumbnail, width=128, height=193, title, authors, status, selected=false){
        this.id = id;
        this.thumbnail = thumbnail;
        this.width = width;
        this.height = height;
        this.title = title;
        this.authors = authors;
        this.status = status;
        this.selected = selected;
    }

}

export default BookModel;