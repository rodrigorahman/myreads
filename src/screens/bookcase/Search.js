import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { search } from '../../BooksAPI'
import PropTypes from 'prop-types'
import Book from './Book'

import Progress from 'react-progress-2'
import BookModel from '../../domain/BookModel'

class Search extends Component {

  static propTypes = {
    books: PropTypes.array.isRequired,
    callbackUpdateState: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired
  }

  state = {
    booksFound: [],
    message: ''
  }

  addBook = (book) => {
    this.props.books.push(book)
    this.props.history.push('/')
  }

  findBooks = (e) => {

    const term = e.target.value

    if (term) {
      Progress.show()
      search(term, 1000)
        .then((data) => {
          let booksFound = []
          console.log(data.error)
          if (!data.error) {

            booksFound = data.map((bf) => {
              const book = this.props.books.find((b) => b.id === bf.id)
              let status = book ? book.status : 'none'
              let authors = bf.authors || []
              let smallImage = (bf.imageLinks ? bf.imageLinks.smallThumbnail : '')

              return new BookModel(
                bf.id,
                smallImage,
                128,
                193,
                bf.title,
                authors,
                status,
                false
              )
            })
            this.setState({ message: '' })
          } else {
            booksFound = []
            this.setState({ message: 'Search did not return result' })
          }

          this.setState({ booksFound })

          Progress.hide()

        }).catch(e => {
        Progress.hide()
        console.error(e)
      })
    }

  }

  render () {

    return (
      <div className="search-books">
        <div className="search-books-bar">
          <Link to='/' className='close-search'>Close</Link>
          <div className="search-books-input-wrapper">
            <input id='txt_term_search' type="text" placeholder="Search by title or author"
                   onChange={ this.findBooks }/>
          </div>
          <Progress.Component/>
        </div>
        <div className="search-books-results">
          <h1>{ this.state.message }</h1>
          <ol className="books-grid">
            {

              this.state.booksFound.map((b) => (
                <li key={ b.id }>
                  <Book
                    callbackUpdateState={ this.props.callbackUpdateState }
                    book={ b }
                    onAddBooking={ this.addBook }
                    multiSelection={ false }
                    history={ this.props.history }
                  />
                </li>
              ))
            }
          </ol>
        </div>
      </div>
    )
  }
}

export default Search