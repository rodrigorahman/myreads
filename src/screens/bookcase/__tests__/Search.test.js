import React from 'react'

import { mount, shallow } from 'enzyme'
import Search from '../Search'
import { BrowserRouter } from 'react-router-dom'
import BookModel from '../../../domain/BookModel'

it('renders without crashing ', () => {

  let books = [new BookModel(1, '', 128, 193, '', ['author 1'], 'none', false)]
  let callbackUpdateState = jest.fn()

  expect(shallow(<Search
    books={ books }
    callbackUpdateState={ callbackUpdateState }
  />)).toMatchSnapshot()
})

it('test with find books', () => {

  let searchResult = [{
    'title': 'Best Android Apps',
    'subtitle': 'The Guide for Discriminating Downloaders',
    'authors': [
      'Mike Hendrickson',
      'Brian Sawyer'
    ],
    'publisher': '"O\'Reilly Media, Inc."',
    'publishedDate': '2010-04-27',
    'description': 'Contains descriptions of over two hundred recommended applications and games for android/mobile devices, including apps for business, communication, lifestyle, entertainment, utility/tool, and reference.',
    'industryIdentifiers': [
      {
        'type': 'ISBN_13',
        'identifier': '9781449382551'
      },
      {
        'type': 'ISBN_10',
        'identifier': '144938255X'
      }
    ],
    'readingModes': {
      'text': false,
      'image': false
    },
    'pageCount': 240,
    'printType': 'BOOK',
    'categories': [
      'Computers'
    ],
    'averageRating': 4,
    'ratingsCount': 3,
    'maturityRating': 'NOT_MATURE',
    'allowAnonLogging': false,
    'contentVersion': 'preview-1.0.0',
    'imageLinks': {
      'smallThumbnail': 'http://books.google.com/books/content?id=bUybAgAAQBAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api',
      'thumbnail': 'http://books.google.com/books/content?id=bUybAgAAQBAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api'
    },
    'language': 'en',
    'previewLink': 'http://books.google.com/books?id=bUybAgAAQBAJ&dq=android&hl=&cd=1&source=gbs_api',
    'infoLink': 'http://books.google.com/books?id=bUybAgAAQBAJ&dq=android&hl=&source=gbs_api',
    'canonicalVolumeLink': 'https://books.google.com/books/about/Best_Android_Apps.html?hl=&id=bUybAgAAQBAJ',
    'id': 'bUybAgAAQBAJ'
  }]

  global.fetch = jest.fn(() => new Promise(resolve => resolve({
    json: () => { return { books: searchResult }}
  })))

  let books = [new BookModel(1, '', 128, 193, '', ['author 1'], 'none', false)]
  let callbackUpdateState = jest.fn()

  let component = mount(
    <BrowserRouter>
      <Search
        books={ books }
        callbackUpdateState={ callbackUpdateState }
      />
    </BrowserRouter>
  )

  let f2 = component.find('#txt_term_search')
  f2.simulate('change', { target: { value: 'a' } })

})

it('test with find books with error', () => {

  global.fetch = jest.fn(() => new Promise(reject => {throw new Error('Oooops!')}))

  let books = [new BookModel(1, '', 128, 193, '', ['author 1'], 'none', false)]
  let callbackUpdateState = jest.fn()

  let component = mount(
    <BrowserRouter>
      <Search
        books={ books }
        callbackUpdateState={ callbackUpdateState }
      />
    </BrowserRouter>
  )

  let f2 = component.find('#txt_term_search')
  f2.simulate('change', { target: { value: 'a' } })

})

it('test with find books and return data.error', () => {

  global.fetch = jest.fn(() => new Promise(resolve => resolve(
    {
      json: () => {
        return {
          books: {
            error: 'empty result'
          }
        }
      }
    })))

  let books = []
  let callbackUpdateState = jest.fn()

  let component = mount(
    <BrowserRouter>
      <Search
        books={ books }
        callbackUpdateState={ callbackUpdateState }
      />
    </BrowserRouter>
  )

  let f2 = component.find('#txt_term_search')
  f2.simulate('change', { target: { value: 'a' } })
})
