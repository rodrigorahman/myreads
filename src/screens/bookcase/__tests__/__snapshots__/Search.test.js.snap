// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing  1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <Link
      className="close-search"
      replace={false}
      to="/"
    >
      Close
    </Link>
    <div
      className="search-books-input-wrapper"
    >
      <input
        id="txt_term_search"
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
      />
    </div>
    <Component
      cls=""
      style={Object {}}
      thumbStyle={Object {}}
    />
  </div>
  <div
    className="search-books-results"
  >
    <h1 />
    <ol
      className="books-grid"
    />
  </div>
</div>
`;
